### PR TITLE
SONARJAVA-1709 Extend list of detected methods for SQL injection

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/SQLInjectionCheck.java [hunk 1]
```diff
@@ -20,63 +20,102 @@
...
 
-  private static boolean isExecuteQueryOrPrepareStatement(MethodInvocationTree methodTree) {
-    return !methodTree.arguments().isEmpty() && (STATEMENT_EXECUTE_QUERY_MATCHER.matches(methodTree) || CONNECTION_MATCHERS.anyMatch(methodTree));
+  @Override
+  public void visitNode(Tree tree) {
+    if (anyMatch(tree)) {
+      Optional<ExpressionTree> sqlStringArg = arguments(tree)
+        .filter(arg -> arg.symbolType().is("java.lang.String"))
+        .findFirst();
+      sqlStringArg.filter(arg -> isDynamicString(tree, arg, null, true)).ifPresent(arg -> {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/SQLInjectionCheck.java]: MINOR Remove useless curly braces around statement [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1602)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: java-checks/src/main/java/org/sonar/java/checks/SQLInjectionCheck.java [hunk 2]
```diff
@@ -20,63 +20,101 @@
...
+    if (methodTree.is(Tree.Kind.NEW_CLASS)) {
+      return ((NewClassTree) methodTree).arguments().stream();
+    }
+    return Stream.empty();
   }
 
-  private static boolean isEntityManagerCreateNativeQuery(MethodInvocationTree methodTree) {
-    return ENTITY_MANAGER_CREATE_NATIVE_QUERY_MATCHER.matches(methodTree);
+  private static boolean anyMatch(Tree tree) {
+    if (!arguments(tree).findAny().isPresent()) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/SQLInjectionCheck.java]: use a dedicated method `hasParameter(tree)`

File: java-checks/src/main/java/org/sonar/java/checks/SQLInjectionCheck.java [hunk 3]
```diff
@@ -20,63 +20,101 @@
...
+    matcherBuilder(SPRING_JDBC_OPERATIONS).name("queryForLong").withAnyParameters(),
+    matcherBuilder(SPRING_JDBC_OPERATIONS).name("update").withAnyParameters(),
+    MethodMatcher.create().typeDefinition("org.springframework.jdbc.core.PreparedStatementCreatorFactory").name("<init>").withAnyParameters(),
+    MethodMatcher.create().typeDefinition("org.springframework.jdbc.core.PreparedStatementCreatorFactory").name("newPreparedStatementCreator").withAnyParameters(),
 
-  private static final MethodMatcher ENTITY_MANAGER_CREATE_NATIVE_QUERY_MATCHER = MethodMatcher.create()
-    .typeDefinition("javax.persistence.EntityManager")
-    .name("createNativeQuery")
-    .withAnyParameters();
+    matcherBuilder("javax.jdo.PersistenceManager").name("newQuery").withAnyParameters(),
```
Comment (saberduck) [inline java-checks/src/main/java/org/sonar/java/checks/SQLInjectionCheck.java]: I updated the RSPEC directly
https://jira.sonarsource.com/browse/RSPEC-2779

**Pull request merged**